### ISSUE TITLE
After upgrading fromSystemJS 0.16 to 0.18 my angular dependencies intermittently fail to load 

### ISSUE THREAD

**Issue by kburson** [systemjs/systemjs#593]
My app will fail to load throwing the following error:
```javascriot
bootstrap.js:30 
a: angular.module is not a function
	Error loading https://localhost:8000/js/components/vertex.js TypeError: angular.module is not a function
    at factory (https://localhost:8000/js/lib/angular-spinner.js:13:5)
    at d.execute (https://localhost:8000/js/lib/system.js:4:28626)
    at i (https://localhost:8000/js/lib/system.js:4:18627)
    at a (https://localhost:8000/js/lib/system.js:4:17276)
    at Object.Promise.all.then.execute (https://localhost:8000/js/lib/system.js:4:21831)
    at y (https://localhost:8000/js/lib/system.js:4:7251)
    at S (https://localhost:8000/js/lib/system.js:4:7630)
    at p (https://localhost:8000/js/lib/system.js:4:5715)
    at h (https://localhost:8000/js/lib/system.js:4:6100)
    at https://localhost:8000/js/lib/system.js:4:4475 
```

my system.config file looks like:

```javascript
System.config({
  baseURL: '/js/components/', //relative to index.html location

  // all paths are relative to baseURL
  paths: {
    'jquery'             : '/js/lib/jquery.min.js',
    'lodash'             : '/js/lib/lodash.js',
    'angular'            : '/js/lib/angular.js',
    'angular-animate'    : '/js/lib/angular-animate.js',
    'angular-messages'   : '/js/lib/angular-messages.js',
    'angular-aria'       : '/js/lib/angular-aria.js',
    'angular-resource'   : '/js/lib/angular-resource.js',
    'angular-cookies'    : '/js/lib/angular-cookies.js',
    'angular-storage'    : '/js/lib/angular-storage.js',
    'angular-material'   : '/js/lib/angular-material.js',
    'angular-ui-router'  : '/js/lib/angular-ui-router.js',
    'angular-sanitize'   : '/js/lib/angular-sanitize.js',
    'angular-mocks'      : '/js/lib/angular-mocks.js',
    'spin'               : '/js/lib/spin.js',
    'angular-spinner'    : '/js/lib/angular-spinner.js',
    'mocks/*.json'       : '/mocks/*.json',
    'l10n/*'             : '/l10n/*'
  },

  // create aliases
  // SystemJS Plugins, referenced with "moduleName!pluginName"
  map: {
    'json': '/js/lib/json.js',
    'text': '/js/lib/text.js'
  },

  // give hints for module dependencies
  meta: {
    'angular'            : {format: 'global', exports: 'angular'},
    'angular-ui-router'  : {format: 'global', deps: ['angular']},
    'angular-animate'    : {format: 'global', deps: ['angular']},
    'angular-aria'       : {format: 'global', deps: ['angular']},
    'angular-cookies'    : {format: 'global', deps: ['angular']},
    'angular-loading-bar': {format: 'global', deps: ['angular']},
    'angular-material'   : {format: 'global', 
        deps: [
            'angular',
            'angular-messages',
            'angular-aria',
            'angular-animate'
        ]},
    'angular-messages'   : {format: 'global', deps: ['angular']},
    'angular-mocks'      : {format: 'global', deps: ['angular']},
    'angular-resource'   : {format: 'global', deps: ['angular']},
    'angular-storage'    : {format: 'global', deps: ['angular']},
    'angular-dropdowns'  : {format: 'global', deps: ['angular']},
    'angular-translate'  : {format: 'global', deps: ['angular']},
    '*.tpl' : {
      loader: 'text'
    },
    '*.json' : {
      loader: 'json'
    }
  }
```

My boostrap.js looks like:

```javascript
System.import('vertex.js').then(
        function (a) {
          console.log('bootstrapping "vertex"');
          angular.element(document).ready(
            function () {
              angular.bootstrap(document, ['vertex']);
            }
          );
        },
        function (a, b, c) {
          console.log('\na:', a.message, a.stack, '\nb:', b, '\nc:', c);
        }
```

my components look like:

```javascript
'use strict';
/*jshint esnext: true */
import 'jquery';
import angular          from 'angular';
import common        from 'common/common.js';
import home             from 'home/home.js';
import research         from 'research/research.js';
import verify              from 'verify/verify.js';
import VertexRouter  from './VertexRouter.js';

export default angular.module('vertex', [
  common.name,
  home.name,
  verify.name,
  research.name
])
  .config(VertexRouter);
```

This was all working correctly with SystemJS 0.16.  During the upgrade to 0.18 I needed to modify all the import statements to include the filename extensions, other than that I did not change anything.

Every 3rd or 4rth refresh the system loads correctly.  That means that it fails 3/4 times!

Please help, I am trying to upgrade to the latest of SystemJS and then migrate from bower/npm over to JSPM, but this is holding me up.

**Comment by guybedford** [comment]
This will be due to an angular module expecting `angular` but not being properly dependent on angular. The way the timings work out, it may often work, but then will hit the case where it loads before angular. I'd suggest filtering out the dependencies with a binary search, to track down with refreshes until you can find the dependency that is at fault, then ensure it is importing angular / has the relevant shim.

**Comment by kburson** [comment]
Thank you for the reply.  

Yes, I found the dependency, angular-spinner.  It is wrapped in an IIFE with some AMD boilerplate at the bottom. 

```javascript
(function (root) {
	'use strict';

	function factory(angular, Spinner) {
              ...
        }

	if (typeof define === 'function' && define.amd) {
		/* AMD module */
		define(['angular', 'spin'], factory);
	} else {
		/* Browser global */
		factory(root.angular);
	}
}(window));
```

 It expects angular to be loaded into global space.  This dependency is not the first file to load, in fact, the first dependency marked to load is jquery followed by angular (reference the post above)

I reverted back to SystemJS 0.16 and everything works fine.  I will pull fresh branch and try again using the transition attribute you created (defaultJSExtensions), see if that works.  I will check back with you later.
